### PR TITLE
Support testing for multi module projects

### PR DESCRIPTION
**Pull request by stuartwdouglas**
Continuous testing will now run tests from all modules, with the option
to configure the old behaviour.

### CODE REVIEW

**Comment by famod** [comment]
I gave this a try and it does execute a test from an upstream module, but it fails with an `AmbiguousResolutionException`:
```
Ambiguous dependencies for type com.github.famod.modmono_quarkus.core.HelloBean and qualifiers [@Default]
        - java member: com.github.famod.modmono_quarkus.core.testlib.HelloBeanTest#underTest
        - declared on CLASS bean [types=[com.github.famod.modmono_quarkus.core.testlib.HelloBeanTest, java.lang.Object], qualifiers=[@Default, @Any], target=com.github.famod.modmono_quarkus.core.testlib.HelloBeanTest]
        - available beans:
                - CLASS bean [types=[com.github.famod.modmono_quarkus.core.HelloBean, java.lang.Object], qualifiers=[@Default, @Any], target=com.github.famod.modmono_quarkus.core.HelloBean]
                - CLASS bean [types=[com.github.famod.modmono_quarkus.core.HelloBean, java.lang.Object], qualifiers=[@Default, @Any], target=com.github.famod.modmono_quarkus.core.HelloBean]
        at io.quarkus.arc.processor.Beans.resolveInjectionPoint(Beans.java:501)
        at io.quarkus.arc.processor.BeanInfo.init(BeanInfo.java:470)
        at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:254)
        ... 13 more
```

The reproducer is here: https://github.com/famod/modmono-quarkus/tree/cont-testing-multi
- ./mvnw clean install
- ./mvnw quarkus:dev -f dist/
- resume testing -> fails

PS: Please ignore the `mocks` profile for now, it's related to #19152.

**Comment by stuartwdouglas** [comment]
This needs some more work, I need to properly resolve an AppModel for each module. I am not sure how practical this will be with the current codebase, it may need to be delayed until after the planned refactoring work.

**Comment by tbvl** [comment]
@stuartwdouglas Sorry for the delay, but now I finally had a little time to try it out. Testing was exploratory. 

**Observations**
- Overall for us it worked fine for us out of the box. Test in dependent modules gets executed, and re-executed on changes. 
- We did not experience CDI related issues as @famod, but I also have not checked for differences in our setup.
- Ressource paths can give issues. Some of our unit tests are using relative paths to test files in the target folder, which works as long as unit tests are executed in isolation within each module, like surefire does. This it is bad practice on our part and something we will fix, but it does give some input to the documentation of the limitations multi module ct feature. 

** Issues **
I saw these two (minor) issues:

1. @ExtendWith not working until added to cp in the module having the Quarkus runtime
We have a test using Mockito in one of the dependee modules. The classloader seems to not include the mockito libraries.
image
It works fine when adding the dependency to mockito to the module where we execute quarkus from.

2. Custom tags does not get recoqnized
We have defined our own custom test tags (including @UnitTest for annotating our test classes for easy filtering and consistency, but it seems Quarkus only picks up test classes where at least one method is annotated directly with @Test or the class is annotated with @QuarkusTest.
image

Hope this input will help, and I hope you will find time to work on it again after the refactoring work. We still find the feature very valuable.

**Comment by stuartwdouglas** [comment]
So issue 1) is the one that I mentioned was a known problem. It might still be worth merging this with it as a documented limitation, possibly with this disabled by default until it is resolved.

2) is a separate issue, can you open a new issue and I will look at fixing it.

**Comment by stuartwdouglas** [comment]
BTW for point 2 I think if you add `org.junit.platform.commons.annotation.Testable` to the annotation it may work.

**Comment by stuartwdouglas** [comment]
I have added a note about the limitation and changed this so it is not longer the default, however I think it is still worth including.

**Comment by tbvl** [comment]
> BTW for point 2 I think if you add `org.junit.platform.commons.annotation.Testable` to the annotation it may work.

Tried this but same result, i.e. it does not get picked up.

> So issue 1) is the one that I mentioned was a known problem. It might still be worth merging this with it as a documented limitation, possibly with this disabled by default until it is resolved.
> 
> 1. is a separate issue, can you open a new issue and I will look at fixing it.
Done and thanks: #19971

**Comment by stuartwdouglas** [comment]
https://github.com/quarkusio/quarkus/pull/20012 will need to be merged before this one.

**Comment by stuartwdouglas** [comment]
This should be ready to merge

**Review by tbvl — APPROVED** [review]

**Review by tbvl — APPROVED** [review]

**Comment by famod** [comment]
> 
> 
> I gave this a try and it does execute a test from an upstream module, but it fails with an `AmbiguousResolutionException`:
> 
> ```
> Ambiguous dependencies for type com.github.famod.modmono_quarkus.core.HelloBean and qualifiers [@Default]
>         - java member: com.github.famod.modmono_quarkus.core.testlib.HelloBeanTest#underTest
>         - declared on CLASS bean [types=[com.github.famod.modmono_quarkus.core.testlib.HelloBeanTest, java.lang.Object], qualifiers=[@Default, @Any], target=com.github.famod.modmono_quarkus.core.testlib.HelloBeanTest]
>         - available beans:
>                 - CLASS bean [types=[com.github.famod.modmono_quarkus.core.HelloBean, java.lang.Object], qualifiers=[@Default, @Any], target=com.github.famod.modmono_quarkus.core.HelloBean]
>                 - CLASS bean [types=[com.github.famod.modmono_quarkus.core.HelloBean, java.lang.Object], qualifiers=[@Default, @Any], target=com.github.famod.modmono_quarkus.core.HelloBean]
>         at io.quarkus.arc.processor.Beans.resolveInjectionPoint(Beans.java:501)
>         at io.quarkus.arc.processor.BeanInfo.init(BeanInfo.java:470)
>         at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:254)
>         ... 13 more
> ```
> 
> The reproducer is here: https://github.com/famod/modmono-quarkus/tree/cont-testing-multi
> 
>     * ./mvnw clean install
> 
>     * ./mvnw quarkus:dev -f dist/
> 
>     * resume testing -> fails
> 
> 
> PS: Please ignore the `mocks` profile for now, it's related to #19152.

⬆️ still fails, I'm afraid...

**Comment by famod** [comment]
@stuartwdouglas I'm now getting the same results, regardless of `quarkus.test.only-test-application-module` being `true` or `false`. Is this expected?

**Comment by famod** [comment]
...but on the plus side there is no exception anymore! 👍

File: docs/src/main/asciidoc/continuous-testing.adoc [hunk 1]
```diff
@@ -150,6 +150,13 @@ mode run `mvn quarkus:test`.
 
 NOTE: The Dev UI is not available when running in continuous testing mode, as this is provided by dev mode.
 
+== Multi Module Projects
+
+Note that continuous testing supports multi-module projects, so tests in modules other than the application can still
+be run when files are changed. The modules that are run can be controlled using config as listed below.
+
+This is enabled by default, and can be disabled via `quarkus.test.only-test-application-module=false`.
```
Comment (famod) [inline docs/src/main/asciidoc/continuous-testing.adoc:158]:
```suggestion
This is enabled by default, and can be disabled via `quarkus.test.only-test-application-module=true`.
```
❔

File: docs/src/main/asciidoc/continuous-testing.adoc [hunk 2]
```diff
@@ -150,6 +150,13 @@ mode run `mvn quarkus:test`.
 
 NOTE: The Dev UI is not available when running in continuous testing mode, as this is provided by dev mode.
 
+== Multi Module Projects
+
+Note that continuous testing supports multi-module projects, so tests in modules other than the application can still
+be run when files are changed. The modules that are run can be controlled using config as listed below.
```
Comment (famod) [inline docs/src/main/asciidoc/continuous-testing.adoc:156]: > The modules that are run can be controlled using config as listed below.

Isn't there just that one toggle? Reads as if I could exclude/include specific modules...?

**Comment by famod** [comment]
> I'm now getting the same results, regardless of `quarkus.test.only-test-application-module` being `true` or `false`. Is this expected?

My bad, I got confused, partially by the changed docs (see https://github.com/quarkusio/quarkus/pull/19793#discussion_r717965516).

File: docs/src/main/asciidoc/continuous-testing.adoc (hunk 2 above)
Comment (famod) [reply docs/src/main/asciidoc/continuous-testing.adoc:156]: Ah, forget about that! This is refererring to the config properties that are rendered automatically. 🤦‍♂️

**Review by famod — COMMENTED** [review]
Played with it a little, looking good! Can't comment much on the code though, looks like you had to extract/restructure things for the new runner, right?

So LGTM, just that documented property seems inverted.

**Comment by aloubyansky** [comment]
Just in case, I tried this change on top of my bootstrap refactoring branch. The Gradle Kotlin test has passed (I guess it may pass in main on the re-run too) but the import sql one failed as well with
````
2021-09-28T19:30:00.8228268Z [ERROR] Tests run: 3, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 25.785 s <<< FAILURE! - in io.quarkus.hibernate.orm.HibernateHotReloadTestCase
2021-09-28T19:30:00.8239525Z [ERROR] io.quarkus.hibernate.orm.HibernateHotReloadTestCase.testImportSqlWithContinuousTesting  Time elapsed: 7.116 s  <<< FAILURE!
2021-09-28T19:30:00.8245788Z org.opentest4j.AssertionFailedError: expected: <0> but was: <1>
2021-09-28T19:30:00.8252851Z 	at org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:55)
2021-09-28T19:30:00.8259080Z 	at org.junit.jupiter.api.AssertionUtils.failNotEqual(AssertionUtils.java:62)
2021-09-28T19:30:00.8269903Z 	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:166)
2021-09-28T19:30:00.8277951Z 	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:161)
2021-09-28T19:30:00.8285761Z 	at org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:628)
2021-09-28T19:30:00.8296312Z 	at io.quarkus.hibernate.orm.HibernateHotReloadTestCase.testImportSqlWithContinuousTesting(HibernateHotReloadTestCase.java:89)
````

**Comment by stuartwdouglas** [comment]
Hopefully all good now.

**Review by famod — APPROVED** [review]
LGTM, also retested it with my small sample project.

Don't know about that one test failure, looks unrelated.

The mvn part in the other job was actually successful but then the run timed out.

**Comment by aloubyansky** [comment]
@stuartwdouglas do you want this to be backported to 2.3? It's too late for the 2.3.0.Final though.

**Comment by stuartwdouglas** [comment]
This is too big a change, let's just leave it for now.

**Pull request merged**